[Text Directionality] Fix directionality for shadow tree

shallawa opened the pull request:
#### b6ee5d981582cff85e203d2403d422a86c53d922
<pre>
[Text Directionality] Fix directionality for shadow tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=277862">https://bugs.webkit.org/show_bug.cgi?id=277862</a>
<a href="https://rdar.apple.com/133549820">rdar://133549820</a>

Reviewed by Ryosuke Niwa.

This patch aligns the TextDirection computations with the specs [1].

Be aware of this difference in this patch:

(a) TextDirection: is an enum which has only two values: LTR and RTL.
(b) TextDirectionState: is enum which is similar to TextDirection but has an
additional values for Auto and Undefined.

Invalidating the Node textDirection and usesTextDirection was fixed in some
places. This fixed most of the WPT failures. But there are still a few failures
left in the test dir-auto-dynamic-changes.window.html.

[1] <a href="https://html.spec.whatwg.org/multipage/dom.html#attr-dir">https://html.spec.whatwg.org/multipage/dom.html#attr-dir</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/dir-auto-dynamic-changes.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/dir-slots-directionality-expected.txt:
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::notifyNodeInsertedIntoTree):
* Source/WebCore/dom/Element.cpp:
(WebCore::parseTextDirectionState):
(WebCore::elementTextDirectionState):
(WebCore::elementHasValidTextDirectionState):
(WebCore::elementHasAutoTextDirectionState):
(WebCore::updateHasDirAutoFlagForSubtree):
(WebCore::updateElementHasDirAutoFlag):
(WebCore::computeTextDirectionFromText):
(WebCore::computeTextDirection):
(WebCore::computeContainedTextAutoDirection):
(WebCore::computeTextDirectionOfSlotElement):
(WebCore::updateEffectiveTextDirectionOfElementAndShadowTree):
(WebCore::updateEffectiveTextDirectionState):
(WebCore::Element::hasValidTextDirectionState const):
(WebCore::Element::hasAutoTextDirectionState const):
(WebCore::Element::computeAutoTextDirection const):
(WebCore::Element::computeTextDirectionIfDirIsAuto const):
(WebCore::Element::updateEffectiveTextDirectionOfDescendants):
(WebCore::Element::updateEffectiveTextDirectionOfAncestors):
(WebCore::Element::updateEffectiveTextDirection):
(WebCore::Element::dirAttributeChanged):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::childrenChanged):
(WebCore::parseTextDirection): Deleted.
(WebCore::isValidDirValue): Deleted.
(WebCore::elementAffectsDirectionality): Deleted.
(WebCore::setHasDirAutoFlagRecursively): Deleted.
(WebCore::Element::hasDirectionAuto const): Deleted.
(WebCore::Element::directionalityIfDirIsAuto const): Deleted.
(WebCore::Element::updateEffectiveDirectionality): Deleted.
(WebCore::Element::adjustDirectionalityIfNeededAfterChildrenChanged): Deleted.
(WebCore::Element::updateTextDirectionalityAfterInputTypeChange): Deleted.
(WebCore::Element::updateEffectiveDirectionalityOfDirAuto): Deleted.
(WebCore::Element::computeDirectionalityFromText const): Deleted.
(WebCore::Element::adjustDirectionalityIfNeededAfterChildAttributeChanged): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didChangeSlot):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::subtreeHasChanged):
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::setValue):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::insertedIntoAncestor):
(WebCore::HTMLSlotElement::didFinishInsertingNode):
* Source/WebCore/html/HTMLSlotElement.h:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::subtreeHasChanged):
(WebCore::HTMLTextAreaElement::setValueCommon):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::directionForFormData const):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::setValue):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchAllRules):
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):
(WebCore::Style::SharingResolver::canShareStyleWithElement const):
(WebCore::Style::elementHasDirectionAuto): Deleted.

Canonical link: <a href="https://commits.webkit.org/282648@main">https://commits.webkit.org/282648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90070bbd213a7abf740b1764516b2895b5941c72

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63807 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14415 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14695 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51416 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9964 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66876 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32033 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13288 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/58638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69524 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7754 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58735 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55334 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58881 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14116 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6447 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40063 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/39806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,325 @@
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementDescendantsAutoTextDirection(const Element& element)
```
This function computes the text direction of element from text content, we should call it that:
e.g. computeTextDirectionFromText
We should place a URL to where this is spec'ed

shallawa replied:
The function was renamed to `computeTextDirectionFromText()` and a URL to the computation algorithm was added to `computeElementTextDirection()`. Unfortunately the sections of the algorithm can't be referenced individually.

shallawa replied:
A link to the specs was added also.

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,325 @@
+static std::optional<TextDirection> computeElementDescendantsAutoTextDirection(const Element& element)
+{
+    for (RefPtr child = element.firstChild(); child; ) {
+        // Specs: Skip bdi, script, style nodes.
```
I don't think these comments are useful.

shallawa replied:
The algorithm is really complex. And I wanted to explicitly differentiate between the parts of the computation which I borrowed from the specs and the parts which I wrote to control the flow of the code. 

Initially I wrote this case in `computeElementTextDirection()` like this:
```
    switch (textDirectionality) {
        ...
    case TextDirectionality::Auto:
        // Specs: Return the auto directionality of element.
        return computeElementAutoTextDirection(element);
        ...
    }
    return std::nullopt;
```
But it turned out to be wrong and it is not aligned with the specs and I had to change it to be like this

```
    switch (textDirectionality) {
        ...
    case TextDirectionality::Auto:
        // Specs: Return the auto directionality of element.
        // Specs: If directionality is null, then return 'ltr'.
        return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
        ...
    }

    return std::nullopt;
```
This change fixed many layout test failures and I wanted to document this change.

rniwa replied:
Yeah but this comment just repeats the code. It literally just skips bid, script, & style elements.

shallawa replied:
The point is not to repeat what the code is doing. The point to document that this code was brought from the specs as is. So if the code changes accidentally, just by reading the code and the comment, we will know that it is not aligned with the specs anymore.

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,325 @@
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeSlotElementAutoTextDirection(const HTMLSlotElement& slotElement)
```
This computes the text direction of slot element so we should call that: computeSlotElementTextDirection

shallawa replied:
This function was renamed to `computeSlotElementTextDirection()`. Similarly `computeTextFormControlElementAutoTextDirection()` was renamed to `computeTextFormControlElementTextDirection()`

shallawa replied:
This is an outdated comment. The functions were renamed to `computeTextDirectionOfSlotElement()` and `computeTextDirectionOfFormControlElement()`.

rniwa commented on Source/WebCore/dom/ElementTextDirection.h:
```diff
@@ -0,0 +1,55 @@
+    LTR,
+    RTL,
+    Auto,
+    Undefined,
```
What is undefined?

shallawa replied:
The specs of [dir](https://html.spec.whatwg.org/multipage/dom.html#attr-dir) attribute defines it as:

> The attribute's [missing value default](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#missing-value-default) and [invalid value default](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#invalid-value-default) are both the undefined state.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,230 +2736,26 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-static void setHasDirAutoFlagRecursively(Node* firstNode, bool flag, Node* lastNode = nullptr)
+bool Element::hasOwnValidAutoTextDirectionality() const
```
This function name is very confusing. Maybe hasAutoTextDirectionality? I don't think "own" and "valid" are adding anything useful here.

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3097,13 +2894,9 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
-        }
-    }
+    RefPtr parent = parentOrShadowHostElement();
+    if (UNLIKELY(selfOrPrecedingNodesAffectDirAuto() || (parent && parent->usesTextDirection())))
```
I think this code was clearer before with the name "effective text direction".

shallawa replied:
I was trying not to use different names. The difference between "text direction" and "effective text direction" is very subtle. I really do not understand what "effective" means in this context. For me "effective" can be used when something is accumulated on top of each other. For example if you store "transparency" for every node in a tree. And the transparency of a node affects the transparency of its children. EffectiveTransparency, for me, will mean the accumulated transparency of a certain node from the root till we reach this node.

I do not see this analogy with the text direction. But I will revert this change.

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3210,9 +3003,9 @@ void Element::removedFromAncestor(RemovalType removalType, ContainerNode& oldPar
-            setUsesEffectiveTextDirection(false);
+    if (UNLIKELY(usesTextDirection()) && !hasValidTextDirectionality()) {
+        if (auto* parent = parentOrShadowHostElement(); !(parent && parent->usesTextDirection()))
+            setUsesTextDirection(false);
```
Ditto about the old name being more clear.

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -328,11 +328,11 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
 
-    TextDirection effectiveTextDirection() const;
-    void setEffectiveTextDirection(TextDirection);
+    TextDirection textDirection() const;
```
I don't think this rename is an improvement.

shallawa replied:
This change was reverted.

rniwa commented on Source/WebCore/html/HTMLSlotElement.cpp:
```diff
@@ -101,6 +104,13 @@ void HTMLSlotElement::attributeChanged(const QualifiedName& name, const AtomStri
     }
 }
 
+void HTMLSlotElement::didFinishInsertingNode()
```
Why do we need to do this in `didFinishInsertingNode`?

rniwa commented on Source/WebCore/dom/ElementTextDirection.h:
```diff
@@ -0,0 +1,55 @@
+class Element;
+
+// https://html.spec.whatwg.org/multipage/dom.html#attr-dir
+enum class TextDirectionality : uint8_t {
```
Sounds like this should be renamed to TextDirectionState?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,326 @@
+        element.invalidateStyleForSubtree();
+}
+
+void updateElementDescendantsTextDirection(Element& element, std::optional<TextDirection> direction, Element* exclude)
```
`updateEffectiveTextDirectionOfDescendants`?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,326 @@
+    return computeElementTextDirection(element, elementTextDirectionality(element));
+}
+
+static void updateElementSelfTextDirection(Element& element, std::optional<TextDirection> direction, Element* exclude = nullptr)
```
This function name is confusing. It says self but then it updates the descendants of its shadow tree.

shallawa replied:
I assumed the shadow tree is something owned and hidden by the node since it is not exposed directly by DOM.

How about `updateElementAndShadowEffectiveTextDirection()`?

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,326 @@
+
+    if (RefPtr shadowRoot = element.shadowRoot()) {
+        for (Ref child : childrenOfType<Element>(*shadowRoot)) {
+            if (child.ptr() == exclude)
```
This `exclude` business is very confusing. Why is this needed?

shallawa replied:
`updateElementTextDirection()` for element _element_ may call `updateElementAncestorTextDirection()` which calls `updateElementTextDirection()` back for the first ancestor _parent_ which has an auto directionality. `updateElementTextDirection()` for element _parent_ will call `updateElementDescendantsTextDirection()`. One of its descents might be _element_. So we may end up updating the textDirection bits for the element we started with _element_. `exclude` will help skipping the initiating element will going up in the tree and then going down.

rniwa commented on Source/WebCore/dom/ElementTextDirection.cpp:
```diff
@@ -0,0 +1,326 @@
+/*
```
Can we not do this refactoring at the same time we're changing the semantics of all these code?

shallawa replied:
The text direction code was moved back to Element.cpp

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,306 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-static bool elementAffectsDirectionality(const Element& element)
+static bool elementHasTextDirectionState(const Element& element)
```
Maybe we should say `elementHasDefinedTextDirectionState` or `elementHasValidTextDirectionState `?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,306 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementContainedTextAutoDirection(const Element& element)
```
Can we add a spec URL here?

rniwa replied:
This function name is also super confusing. Maybe spec has some name we can adopt here?

rniwa replied:
This is https://html.spec.whatwg.org/#contained-text-auto-directionality, right? We should just call it computeContainedTextAutoDirectionality and add a link here.

shallawa replied:
I got this name from the specs: https://html.spec.whatwg.org/multipage/dom.html#contained-text-auto-directionality.

rniwa replied:
I don't think the extra "element" is adding any value here.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-        if (node->hasTagName(bdiTag) || node->hasTagName(scriptTag) || node->hasTagName(styleTag)
-            || (element && element->isTextField())) {
-            node = NodeTraversal::nextSkippingChildren(*node, this);
+std::optional<TextDirection> Element::computeAutoTextDirection() const
```
This is computing TextDirection when dir=auto is set. We should call it something like `computeTextDirectionForDirAuto` or something.

shallawa replied:
This is not necessarily the case. The specs says: to compute the text direction in the case of dir _undefined_:

> [undefined](https://html.spec.whatwg.org/multipage/dom.html#attr-dir-undefined-state)
>> If element is a [bdi](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-bdi-element) element
>>> Let result be the [auto directionality](https://html.spec.whatwg.org/multipage/dom.html#auto-directionality) of element.

rniwa replied:
Okay, in that case, we should probably call it something like `computeTextDirectionForAutoTextDirectionState`.

rniwa replied:
Hm... maybe we can just refer to https://html.spec.whatwg.org/multipage/dom.html#auto-directionality and call this "computeAutoDirectionality" to match the spec language?

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-                continue;
-            }
-        }
+        if (child.ptr() != exclude)
```
This "exclude" business is very confusing. We don't process the excluded element but we process its descendants?

shallawa replied:
> This "exclude" business is very confusing

Since we update the descendants and the ancestors of element _element_, we have to protect _element_ itself from being reset when updating the descendants any of the _element_'s ancestors.

> We don't process the excluded element but we process its descendants?

You are right. This was a bug. I fixed it by calling `traverseNextSkippingChildren()` when we revisit _element_.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    for (Ref ancestor : lineageOfType<Element>(*this)) {
+        if (!elementHasAutoTextDirectionState(ancestor))
+            continue;
+        updateElementTextDirection(ancestor, TextDirectionState::Auto, exclude);
```
It's very confusing to have two functions named `updateElementTextDirection` and `updateTextDirection`. What's the difference between two? We should make it clear in the function name.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-void Element::updateEffectiveDirectionality(std::optional<TextDirection> direction)
+static std::optional<TextDirection> computeSlotElementTextDirection(const HTMLSlotElement& slotElement)
```
I'd rather call this `computeTextDirectionOfSlotElement`.

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildrenChanged(Element* beforeChange, ChildChange::Type changeType)
+static std::optional<TextDirection> computeTextFormControlElementTextDirection(const HTMLTextFormControlElement& textFormControlElement)
```
And `computeTextDirectionOfFormControlElement`

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildAttributeChanged(Element*)
+void Element::updateEffectiveTextDirectionOfAncestor(Element* exclude)
```
updateEffectiveTextDirectionOfAncestor*s*?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
     }
 }
 
+void Element::updateTextDirection(std::optional<AtomString> newValue)
```
I'd call this `updateEffectiveTextDirection` since that's what this function does.

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 
+void Element::updateTextDirection(std::optional<AtomString> newValue)
+{
+    auto value = newValue.value_or(attributeWithoutSynchronization(dirAttr));
```
This is a very confusing function interface. I think it's better if the caller always passed in newValue instead of conditionally getting the attribute value like this inside the function.

shallawa replied:
Done.

I put back the original function `Element::dirAttributeChanged()`. So I was able to remove the `newValue` argument form this function.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-static void setHasDirAutoFlagRecursively(Node* firstNode, bool flag, Node* lastNode = nullptr)
+static void setNodeHasDirAutoFlagRecursively(Node& firstNode, bool flag)
```
Let's introduce an enum class instead of using a bool. We should also clarify which way this recursion goes (ancestors vs. descendants).
How about `updateHasDirAutoFlagForSubtree`?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-void Element::updateTextDirectionalityAfterInputTypeChange()
+static void updateElementTextDirection(Element& element, TextDirectionState textDirectionState, Element* exclude = nullptr)
```
Maybe we should call this `updateEffectiveTextDirection`?

shallawa replied:
I called it `updateEffectiveTextDirectionState()` since it takes `TextDirectionState`. This is to differentiate it from the functions which take `std::optional<TextDirection>`

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    case TextDirectionState::LTR:
+    case TextDirectionState::RTL:
+        if (element.selfOrPrecedingNodesAffectDirAuto() || (parent && parent->selfOrPrecedingNodesAffectDirAuto()))
+            updateHasDirAutoFlagForSubtree(element, false);
```
Can we define an enum class instead of using boolean here? It's unclear what "false" and "true" means in this context.

shallawa replied:
Instead of creating a new enum, I am passing `textDirectionState` instead and set the selfOrPrecedingNodesAffectDirAuto flag to true only when the `textDirectionState != TextDirectionState::Auto`

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+        // Specs: If child is a slot element whose root is a shadow root,
+        // then return the directionality of that shadow root's host.
+        if (auto* childSlotElement = dynamicDowncast<HTMLSlotElement>(*child)) {
+            if (auto childHost = childSlotElement->shadowHost())
```
Can we use `RefPtr` here?

shallawa replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 
-void Element::adjustDirectionalityIfNeededAfterChildrenChanged(Element* beforeChange, ChildChange::Type changeType)
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+static std::optional<TextDirection> computeTextDirectionOfFormControlElement(const HTMLTextFormControlElement& textFormControlElement)
```
I think should just merge this back into `computeAutoTextDirection() const`. It's only used once in that function and I don't see much value in splitting it.

shallawa replied:
Merged.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-    RefPtr<Node> oldMarkedNode;
-    if (beforeChange)
-        oldMarkedNode = changeType == ChildChange::Type::ElementInserted ? ElementTraversal::nextSibling(*beforeChange) : beforeChange->nextSibling();
+static void updateElementAndShadowEffectiveTextDirection(Element& element, std::optional<TextDirection> direction, Element* initiator = nullptr)
```
I'd rather call this `updateEffectiveTextDirectionOfElementAndShadowTree`.

shallawa replied:
Done.

shallawa committed:
[Text Directionality] Fix directionality for shadow tree
https://bugs.webkit.org/show_bug.cgi?id=277862
rdar://133549820

Reviewed by Ryosuke Niwa.

This patch aligns the TextDirection computations with the specs [1].

Be aware of this difference in this patch:

(a) TextDirection: is an enum which has only two values: LTR and RTL.
(b) TextDirectionState: is enum which is similar to TextDirection but has an
additional values for Auto and Undefined.

Invalidating the Node textDirection and usesTextDirection was fixed in some
places. This fixed most of the WPT failures. But there are still a few failures
left in the test dir-auto-dynamic-changes.window.html.

[1] https://html.spec.whatwg.org/multipage/dom.html#attr-dir

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/dir-auto-dynamic-changes.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/dir-slots-directionality-expected.txt:
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::notifyNodeInsertedIntoTree):
* Source/WebCore/dom/Element.cpp:
(WebCore::parseTextDirectionState):
(WebCore::elementTextDirectionState):
(WebCore::elementHasValidTextDirectionState):
(WebCore::elementHasAutoTextDirectionState):
(WebCore::updateHasDirAutoFlagForSubtree):
(WebCore::updateElementHasDirAutoFlag):
(WebCore::computeTextDirectionFromText):
(WebCore::computeTextDirection):
(WebCore::computeContainedTextAutoDirection):
(WebCore::computeTextDirectionOfSlotElement):
(WebCore::updateEffectiveTextDirectionOfElementAndShadowTree):
(WebCore::updateEffectiveTextDirectionState):
(WebCore::Element::hasValidTextDirectionState const):
(WebCore::Element::hasAutoTextDirectionState const):
(WebCore::Element::computeAutoTextDirection const):
(WebCore::Element::computeTextDirectionIfDirIsAuto const):
(WebCore::Element::updateEffectiveTextDirectionOfDescendants):
(WebCore::Element::updateEffectiveTextDirectionOfAncestors):
(WebCore::Element::updateEffectiveTextDirection):
(WebCore::Element::dirAttributeChanged):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::childrenChanged):
(WebCore::parseTextDirection): Deleted.
(WebCore::isValidDirValue): Deleted.
(WebCore::elementAffectsDirectionality): Deleted.
(WebCore::setHasDirAutoFlagRecursively): Deleted.
(WebCore::Element::hasDirectionAuto const): Deleted.
(WebCore::Element::directionalityIfDirIsAuto const): Deleted.
(WebCore::Element::updateEffectiveDirectionality): Deleted.
(WebCore::Element::adjustDirectionalityIfNeededAfterChildrenChanged): Deleted.
(WebCore::Element::updateTextDirectionalityAfterInputTypeChange): Deleted.
(WebCore::Element::updateEffectiveDirectionalityOfDirAuto): Deleted.
(WebCore::Element::computeDirectionalityFromText const): Deleted.
(WebCore::Element::adjustDirectionalityIfNeededAfterChildAttributeChanged): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didChangeSlot):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::subtreeHasChanged):
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::setValue):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::insertedIntoAncestor):
(WebCore::HTMLSlotElement::didFinishInsertingNode):
* Source/WebCore/html/HTMLSlotElement.h:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::subtreeHasChanged):
(WebCore::HTMLTextAreaElement::setValueCommon):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::directionForFormData const):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::setValue):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchAllRules):
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):
(WebCore::Style::SharingResolver::canShareStyleWithElement const):
(WebCore::Style::elementHasDirectionAuto): Deleted.

Canonical link: https://commits.webkit.org/282648@main
